### ISSUE TITLE
Several issues with process_perfdata.pl

### ISSUE THREAD

**Issue by tiahino** [pnp4nagios/pnp4nagios#56]
This is not your usual bug report - the problems are described and possible solutions offered. I would leave to maintainers to make a final decision.

So the saga is as follows. 

During the deployment (v 0.6.27-4l) I let the pnn4nagios spool directory to grow (to abot 2.5K files). This is usually not a problem but to my surprise the processing speed was very slow - it was barely keeping up with newly generated files. I have increased the number of npcd threads (npcd.conf:npcd_max_threads) from 5 to 10 and reduced the wait time (npcd.conf:sleep_time) from 15 to 10. This actually made things even worse and enabled debug revealed that npcd gets a lot of children (process_perfdata.pl) exiting with status 2:
```
[02-15-2024 17:17:34] NPCD: ERROR: Executed command exits with return code '2'
[02-15-2024 17:17:34] NPCD: ERROR: Command line was '/usr/libexec/pnp4nagios/process_perfdata.pl -n --bulk /dev/shm/pnp4nagios/spool//service-perfdata.1708017412'
```
Check through the source code revealed only one place where it exits deliberately with code 2 and it was related to gearmin mode - i.e. absolutely irrelevant to the case at hand.
The exerpt from perfdata log for the process at hand is even more puzzling:
```
...
2024-02-15 17:17:19 [3631184] [1] /dev/shm/pnp4nagios/spool//service-perfdata.1708017412-PID-3631184 deleted
2024-02-15 17:17:19 [3631184] [2] INFO: did not get stats lock, 60 attempts remain
2024-02-15 17:17:20 [3631184] [2] INFO: did not get stats lock, 59 attempts remain
2024-02-15 17:17:21 [3631184] [2] INFO: did not get stats lock, 58 attempts remain
2024-02-15 17:17:22 [3631184] [2] INFO: did not get stats lock, 57 attempts remain
2024-02-15 17:17:23 [3631184] [2] INFO: did not get stats lock, 56 attempts remain
2024-02-15 17:17:24 [3631184] [2] INFO: did not get stats lock, 55 attempts remain
2024-02-15 17:17:25 [3631184] [2] INFO: did not get stats lock, 54 attempts remain
2024-02-15 17:17:26 [3631184] [2] INFO: did not get stats lock, 53 attempts remain
2024-02-15 17:17:27 [3631184] [2] INFO: did not get stats lock, 52 attempts remain
2024-02-15 17:17:28 [3631184] [2] INFO: did not get stats lock, 51 attempts remain
2024-02-15 17:17:29 [3631184] [2] INFO: did not get stats lock, 50 attempts remain
2024-02-15 17:17:30 [3631184] [2] INFO: did not get stats lock, 49 attempts remain
2024-02-15 17:17:31 [3631184] [2] INFO: did not get stats lock, 48 attempts remain
2024-02-15 17:17:32 [3631184] [2] INFO: did not get stats lock, 47 attempts remain
2024-02-15 17:17:33 [3631184] [2] INFO: did not get stats lock, 46 attempts remain
2024-02-15 17:17:34 [3631184] [0] *** TIMEOUT: Timeout after 15 secs. ***
2024-02-15 17:17:34 [3631184] [0] *** TIMEOUT: Deleting current file to avoid NPCD loops
2024-02-15 17:17:34 [3631184] [0] *** TIMEOUT: Please check your process_perfdata.cfg
2024-02-15 17:17:34 [3631184] [0] *** TIMEOUT: Could not delete /dev/shm/pnp4nagios/spool//service-perfdata.1708017412-PID-3631184:No such file or directory
```
The puzzling part is 1) why does it have any errors deleting the file (after reporting it deleted successfully) and 2) all the exit statements in the code have some reason printed but the above set looks like abruptly terminated. And 3) why does it get into the ALARM handler at all?

The program sets the alarm for 15 seconds before it goes off to the races (process a single file passed as an argument) - and in my case it actually completes the work in one second. And before it is completely done it does some stats reporting:
```perl
    220     $rt = tv_interval($t0, [gettimeofday]);
    221     $stats{runtime} = $rt;
    222     $stats{rows} = $lines;
    223     store_internals();
    224     print_log( "PNP exiting (runtime ${rt}s) ...", 1 );
    225     exit 0;
```

The above is the normal sequence and one would expect the exit 0. However the store_internals() tries to process accumulated stats and put them into a special rrd database. In 0.6.27 an attempt was made to serialize this operation and a locking mechanism was introduced:
```perl
   1364 sub store_internals {
   1365     if( ! -w $conf{'STATS_DIR'}){
   1366         print_log("*** ERROR: ".$conf{'STATS_DIR'}." is not writable or does not exist",0);
   1367         return;
   1368     }
   1369     my $lockfile = $conf{'STATS_DIR'}."/".'stats_lockfile';
   1370     open(LOCK, ">>$lockfile") or die "cannot create stats lock file ",$!;
   1371     my $lockstat;
   1372     my $locktries = 60;
   1373     while ($locktries > 0) {
   1374         $lockstat  = flock(LOCK, LOCK_EX|LOCK_NB);
   1375         last if $lockstat;
   1376         print_log("INFO: did not get stats lock, $locktries attempts remain",2);
   1377         $locktries--;
   1378         sleep(1) if $locktries > 0;
   1379     }
   1380     if (!$lockstat) {
   1381         print_log("ERROR: failed to get stats file lock, bailing",0);
   1382         return;
   1383     } else {
   1384     print_log("DEBUG: got stats file lock",3);
   1385     }
   1386
   1387     my $statsfile = $conf{'STATS_DIR'}."/".(int $stats{timet} / 60);
   1388     open( STAT, ">> $statsfile" ) or die "Cant create statistic file ", $!;
   1389     printf(STAT "%d %f %d %d %d %d %d %d\n", $stats{timet},$stats{runtime},$stats{rows},$stats{update},$stats{create},$stats{error},$stats{invalid},$stats{skipped});
   1390     close(STAT);
   1391     sleep(2);  # prevent rrd update too soon?
   1392     check_internals();
   1393     close(LOCK);
   1394 }
```
Note that the process tries to write the stats into the file which is unique down to **minute** (int $stats{timet} / 60) . Later on, in check_internals() it will consume all accumulated stats files **except the one it just wrote** which means that until the minute is over, the current file won't get processed. This is important to understand as the store_internals() implementation above holds the lock for two seconds for no reason - preventing other processes from getting the lock and continuing. Now imaging we have ten processing trying to get through the process - and each process only has 15 seconds before it is killed - which gives us only about 3-4 processes to have chance to terminate gracefully.

Hence my first recommendation - the line 1391 above should be dropped as it does not provide any additional protection and only slows down things considerably. To put it another way - within 15 seconds the stats file would be written once (maybe twice if you are right on the minute border) so making other processes wait for even a second is a waste of time. We could make it faster if we replace sleep with usleep as we use Time::HiRes anyway, i.e.:
```perl
use Time::HiRes qw(usleep);
...
usleep(500000) if $locktries > 0; # Sleep for 500 milliseconds (0.5 seconds)
```

Now let's see what happens when the SIGALARM is received. Remember, the signal is received when the process already completed processing the input file, deleted it and is waiting to update the stats:
```perl
   1313         if ( $signal eq "ALRM" ) {
   1314             print_log( "*** TIMEOUT: Timeout after $opt_t secs. ***", 0 );
   1315             if ( $opt_b && !$opt_n && !$opt_s ) {
   1316                 print_log( "*** TIMEOUT: Deleting current file to avoid loops",   0 );
   1317                 print_log( "*** TIMEOUT: Please check your process_perfdata.cfg", 0 );
   1318             }
   1319             elsif ( $opt_b && $opt_n && !$opt_s ) {
   1320                 print_log( "*** TIMEOUT: Deleting current file to avoid NPCD loops", 0 );
   1321                 print_log( "*** TIMEOUT: Please check your process_perfdata.cfg",    0 );
   1322             }
   1323             if ($opt_b && !$opt_s ) {
   1324                 my $pdfile = "$opt_b" . "-PID-" . $$;
   1325                 if ( unlink("$pdfile") == 1 ) {
   1326                     print_log( "*** TIMEOUT: $pdfile deleted", 0 );
   1327                 }
   1328                 else {
   1329                     print_log( "*** TIMEOUT: Could not delete $pdfile:$!", 0 );
   1330                 }
   1331             }
   1332             my $temp_file = "$conf{RRDPATH}/$NAGIOS{HOSTNAME}/$NAGIOS{SERVICEDESC}.xml.$$";
   1333             if ( -f $temp_file ) {
   1334                 unlink($temp_file);
   1335             }
   1336             $t1 = [gettimeofday];
   1337             $rt = tv_interval $t0, $t1;
   1338             $stats{runtime} = $rt;
   1339             print_log( "*** Timeout while processing Host: \"$NAGIOS{HOSTNAME}\" Service: \"$NAGIOS{SERVICEDESC}\"", 0 );
   1340             print_log( "*** process_perfdata.pl terminated on signal $signal", 0 );
   1341             exit 7;
   1342         }
```
Note lines 1324-1331. The printout from 1329 is what I was seeing in the logs. I think it should better be written this way:
```perl
            if ($opt_b && !$opt_s ) {
                my $pdfile = "$opt_b" . "-PID-" . $$;
                if ( -f $pdfile ) {
                    if ( unlink("$pdfile") == 1 ) {
                        print_log( "*** TIMEOUT: $pdfile deleted", 0 );
                    }
                    else {
                        print_log( "*** TIMEOUT: Could not delete $pdfile:$!", 0 );
                    }
                }
            }
```
Now if we're getting into this handler we should have exited with code 7 and have seen the messages from lines 1339 and 1340 in the log. This does not happen and took me a while to ascertain that tv_interval on the line 1337 is the real black hole where the process just crashes. Look at the parameters for it - they are supposed to be array references, a.k.a. pointers. $t1 is initialized right there, but what about $t0? Turns out that $t0 is declared as a global variable at the beginning of the file
```perl
use vars qw ( $TEMPLATE %NAGIOS $t0 $t1 $rt $delayed_write $rrdfile @ds_create $count $line $name $ds_update $dstype %CTPL);
```
but what is initialized in main() is a local variable:
```perl
sub main {
    my $job = shift;
    my $t0 = [gettimeofday];
		...
```
so in the signal handler $t0 is uninitalized pointer. And we all know what happens when uninitialized pointers are used.
Changing the assignment in main() like this:
```perl
		$t0 = [gettimeofday]
```
fixes this last problem and with all three fixes (the sleep removal, check for file existence and initializing the reference point $t0) the process behaves normally, producing expected log records. Actually because the processes no longer have to wait extra 2 seconds and all 10 threads can finish within allotted 15 seconds.

Please consider integrating the above fixes into the source.

**Comment by celane** [comment]
All good diagnosis and suggestions, and I'll be implementing those changes. 
And probably a few more.

The difficulty is that nagios is generating performance data asynchronously, and then the processing occurs; perhaps on a regular interval with npcd,  but in a less regular fashion with bulk and synchronous modes.  Resulting in errors when an earlier "bulk" processing run handles the file that a later  processing run was expecting to find.  I'm tempted to just comment out the error messages, but there are actual error modes that would then be hidden (Nagios writes to directory A, processing expects it's in directory B, etc). 

Then there's the complaints from RRD if it gets too many updates too fast.  That's maybe a "startup transient" problem, because Nagios doesn't generate perfdata super-fast, but I've still seen it happen.  

I *am* tempted to move to using "rrdcached", which would at least make the connection between process_perfdata and rrd  more efficient. Maybe in the next iteration.

**Comment by celane** [comment]
changes in locking and file handling, in testing branch, now merged into 0.6.27-5; to improve efficiency (more granular and less distro-specific locking). So marking this issue closed. Please open an issue if there are still performance improvements that should be made.
